commit 282e4ae275dc1b9b0b5bba6eb1b145cd1e80fa33
Author: Justus Winter <4winter@informatik.uni-hamburg.de>
Date:   Fri Mar 14 00:29:02 2014 +0100

    mach-defpager: use protected payloads for object lookups
    
    * mach-defpager/default_pager.c (pager_port_list_insert): Set
    protected payload.
    (pager_port_list_delete): Clear protected payload.
    * mach-defpager/mig-decls.h (begin_using_default_pager_payload): New
    function.
    * mach-defpager/mig-mutate.h: Add mutator.

diff --git b/mach-defpager/default_pager.c a/mach-defpager/default_pager.c
index 831ed96..d40a9c9 100644
--- b/mach-defpager/default_pager.c
+++ a/mach-defpager/default_pager.c
@@ -1803,12 +1803,6 @@ void pager_port_list_insert(port, ds)
 			(hurd_ihash_key_t) port,
 			(hurd_ihash_value_t) ds);
 	pthread_mutex_unlock(&all_pagers.lock);
-
-	/* Try to set a protected payload.  This is an optimization,
-	   if it fails we degrade gracefully.  */
-	mach_port_set_protected_payload (mach_task_self (),
-					 port,
-					 (unsigned long) ds);
 }
 
 void pager_port_list_delete(ds)
@@ -1818,9 +1812,6 @@ void pager_port_list_delete(ds)
 	hurd_ihash_locp_remove (&all_pagers.htable,
 				ds->htable_locp);
 	pthread_mutex_unlock(&all_pagers.lock);
-
-	mach_port_clear_protected_payload (mach_task_self (),
-					   ds->pager);
 }
 
 /*
diff --git b/mach-defpager/mig-decls.h a/mach-defpager/mig-decls.h
index 3357aa2..8118d61 100644
--- b/mach-defpager/mig-decls.h
+++ a/mach-defpager/mig-decls.h
@@ -31,10 +31,4 @@ begin_using_default_pager (mach_port_t port)
                                             (hurd_ihash_key_t) port);
 }
 
-static inline struct dstruct * __attribute__ ((unused))
-begin_using_default_pager_payload (unsigned long payload)
-{
-  return (default_pager_t) payload;
-}
-
 #endif /* __MACH_DEFPAGER_MIG_DECLS_H__ */
diff --git b/mach-defpager/mig-mutate.h a/mach-defpager/mig-mutate.h
index 31acaa6..54aeeba 100644
--- b/mach-defpager/mig-mutate.h
+++ a/mach-defpager/mig-mutate.h
@@ -18,8 +18,5 @@
    along with the GNU Hurd.  If not, see <http://www.gnu.org/licenses/>.  */
 
 #define MEMORY_OBJECT_INTRAN default_pager_t begin_using_default_pager (mach_port_t)
-#define MEMORY_OBJECT_INTRAN_PAYLOAD			\
-  default_pager_t begin_using_default_pager_payload
-
 #define MEMORY_OBJECT_IMPORTS import "mig-decls.h";
 #define DEFAULT_PAGER_IMPORTS import "mig-decls.h";
